### PR TITLE
Make auth argument in the register request compliant with r0.6.0

### PR DESCRIPTION
**Pull request by MTRNord**
This fix has a companion PR over at react-sdk. https://github.com/matrix-org/matrix-react-sdk/pull/4347

This fix is tested with synapse and works fine on matrix.org!

This fix is required for https://git.koesters.xyz/timo/matrixserver which implements the r0.6.0 spec. In that spec version it says that auth shouldn't be `auth: {}` but rather not be present. This fix should fix this. I tested it both with synapse and the linked server to check that everything still works

Signed-off-by: Marcel Radzio<mtrnord1@gmail.com>

### CODE REVIEW

File: src/interactive-auth.js [hunk 1]
```diff
@@ -318,14 +318,18 @@ InteractiveAuth.prototype = {
      */
     _doRequest: async function(auth, background) {
         try {
+            if (Object.keys(auth).length === 0 &&
+                auth.constructor === Object) auth = null;
+            console.warn(auth);
```
Comment (turt2live) [inline src/interactive-auth.js:323]: This debugging should be removed

File: src/interactive-auth.js [hunk 2]
```diff
@@ -162,14 +162,14 @@ InteractiveAuth.prototype = {
...
-        if (!this._data.session) return;
+        if (this._data && !this._data.session) return;
         // likewise don't poll if there is no auth session in progress
         if (!this._resolveFunc) return;
         // if we currently have a request in flight, there's no point making
         // another just to check what the status is
         if (this._submitPromise) return;
 
-        let authDict = {};
+        let authDict = null;
```
Comment (turt2live) [inline src/interactive-auth.js:172]: This is supposed to be something by here

File: src/interactive-auth.js [hunk 3]
```diff
@@ -162,14 +162,14 @@ InteractiveAuth.prototype = {
      * be resolved.
      */
     poll: async function() {
-        if (!this._data.session) return;
+        if (this._data && !this._data.session) return;
```
Comment (turt2live) [inline src/interactive-auth.js:165]: we should have `this._data`

File: src/interactive-auth.js [hunk 4]
```diff
@@ -318,14 +318,18 @@ InteractiveAuth.prototype = {
      */
     _doRequest: async function(auth, background) {
         try {
+            if (Object.keys(auth).length === 0 &&
+                auth.constructor === Object) auth = null;
```
Comment (turt2live) [inline src/interactive-auth.js:322]: I don't think we need to modify `auth` here - we should be supplying the function with what it needs.

File: src/interactive-auth.js [hunk 5]
```diff
@@ -318,14 +318,18 @@ InteractiveAuth.prototype = {
...
+            console.warn(auth);
             const result = await this._requestCallback(auth, background);
             this._resolveFunc(result);
             this._resolveFunc = null;
             this._rejectFunc = null;
         } catch (error) {
             // sometimes UI auth errors don't come with flows
             const errorFlows = error.data ? error.data.flows : null;
-            const haveFlows = Boolean(this._data.flows) || Boolean(errorFlows);
+            const haveFlows = Boolean(this._data &&
```
Comment (turt2live) [inline src/interactive-auth.js:331]: we don't need this first `Boolean` cast anymore

File: src/interactive-auth.js [hunk 6]
```diff
@@ -368,7 +372,7 @@ InteractiveAuth.prototype = {
                         this._inputs.emailAddress,
                         this._clientSecret,
                         1, // TODO: Multiple send attempts?
-                        this._data.session,
+                        this._data ? this._data.session : null,
```
Comment (turt2live) [inline src/interactive-auth.js:375]: we should have a session by here

File: src/interactive-auth.js (hunk 1 above)
Comment (MTRNord) [reply src/interactive-auth.js:323]: Oh sorry missed that one. Will do that!

File: src/interactive-auth.js (hunk 4 above)
Comment (MTRNord) [reply src/interactive-auth.js:322]: I think you are right. This seems to be a bad workaround that I made. I should do it properly as I probably now broke things without realising.

**Comment by turt2live** [comment]
@MTRNord is this ready for re-review?

**Comment by MTRNord** [comment]
@turt2live No I didn't have time today to work on this. I will work on this tomorrow. (German timezone)

**Comment by MTRNord** [comment]
@turt2live I think this is fixed now. The real change is made in the FallbackStage on the react-sdk side. I think this should be everything. Currently building the docker for another local test

**Comment by MTRNord** [comment]
Actually no thats not the correct stage I changed. I am confused :/ I tracked it back to the InteractiveAuthEntryComponents.js in react-sdk but I am not sure which part gets triggered with the "m.login.dummy" stage

**Comment by MTRNord** [comment]
@turt2live I give up. I cant get this to work anymore. Not sure where I need to change something to have that initial stage with auth: null instead auth: {} in the code.

**Comment by turt2live** [comment]
the system is quite complicated, sorry :(

I'll try and take a look next week

**Comment by turt2live** [comment]
I'm not going to be able to take a look, sorry :(

Leaving this in the pile of things people are welcome to try and help with.

**Comment by turt2live** [comment]
right, this should be fixed and a whole lot of reruns later the CI is happy. Asking for review from someone else on the team given I've put some code into this now.

**Comment by MTRNord** [comment]
Thanks for the work on this to make it actually work!

**Comment by jryans** [comment]
(Merged now as this was our last release blocker, and I'd like to start the RC process.)

**Pull request merged**